### PR TITLE
iptables proxier: route local traffic to LB IPs to service chain

### LINKED ISSUE #65387: externalTrafficPolicy: Local breaks internal reachability
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

This is discussed on [kubernetes-users](https://groups.google.com/forum/#!topic/kubernetes-users/HOaGQD4J8j0)

When externalTrafficPolicy: Local is used, in GKE, external traffic coming in to the Load Balancer works correctly. But traffic that originates within the GKE cluster is broken, a TCP SYN is not responded to.

My use case:
I run gitlab, gitlab-runner, docker-registry, wiki inside kubernetes.

But I also access the docker-registry outside kubernetes. So my hostname for it is 'cr.COMPANY.com'. In DNS this resolves to the LB IP. Actually, all 3 services have DNS names that resolve to the same LB IP, which then goes to the nginx ingress. so wiki.COMPANY.com, git.COMPANY.com, wiki.COMPANY.com are generally addressed as LB-IP w/ HOSTNAME+SNI. Its 'astonishing' to not be able to access this DNS name when I'm 'inside' but can when I'm 'outside'.

If I enable the IP transparency (because in my wiki I want the IP of the user who made the edit), then my gitlab-runner fails(since it cannot pull from the registry).

Tim Hockin has this to say:

```
On Friday, June 22, 2018 at 1:32:45 PM UTC-4, Tim Hockin wrote:
I reproduced this.  Here's the explanation of why it is working as intended.  We can discuss whether that intention is misguided or not.

When externalTrafficPolicy is "Cluster", each node acts as an LB gateway to the Service, regardless of where the backends might be running.  In order to make this work, we must SNAT (which obscures the client's IP) because the traffic could cross nodes, and you can't have the final node responding directly to the client (bad 5-tuple).  As we agree, this mode works but isn't what you want.

When externalTrafficPolicy is "Local", only nodes that actually have a backend for a given Service act as an LB gateway.  This means we do not need to SNAT, thereby keeping the client IP.  But what about nodes which do not have backends?  They drop those packets.  The GCE load-balancer has relatively long programming time, so we don't want to change the TargetPools every time a pod moves.  The healthchecks are set up such that nodes that do not have a backend fail the HC - thus the LB should never route to them!  Very clever.

Here's the rub - the way GCP sets up LBs is via a VIP which is a local IP to each VM.  By default, access to the LB VIP from a node "behind" that VIP (all nodes in k8s) is serviced by that same VM, not by the actual LB.  The assumption is that you are accessing yourself, why go through the network to do that?

Kubernetes makes an explicit provision for pods that access an LB VIP by treating them as if they accessed the internal service VIP (which is not guaranteed to stay node-local).  We did not make a provision for a NODE to access the LB VIP in the same way.  Maybe we could?  I seem to recall an issue there, in how we distinguish traffic originating from "this VM" vs traffic we are gatewaying.

So there you see - it is doing what is intended, but maybe not what you want.  Now - concvince me that the use-case of accessing an external LB VIP from a node in the cluster (not a pod - a node) is worth the extra complexity?  One case I admit that falls in a crack is `hostNetwork` pods.  They will fail.
```
and then:

```
On Friday, June 22, 2018 at 4:14:24 PM UTC-4, Tim Hockin wrote:
Well, the good news is that I think this is easy to fix.

For my test case, the service hashes to NWV5X2332I4OT4T3

iptables -t nat -I KUBE-XLB-NWV5X2332I4OT4T3 -m addrtype --src-type LOCAL -j KUBE-SVC-NWV5X2332I4OT4T3
iptables -t nat -I KUBE-XLB-NWV5X2332I4OT4T3 -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ

Note these two are in inverse order because I am lazy and used -I.  This will break the "only local" property for in-cluster accesses, but I think that is OK since it's explicitly not true for access from pods.

The kube-proxy change here should be easy enough, but the testing is a little involved.

Would you do me the good favor of opening a github bug and we can see if we can rally an implementor?
```

**What you expected to happen**:

My expectation is that it should not matter where my traffic originates, if it is to a globally-routeable public IP it should work.

**How to reproduce it (as minimally and precisely as possible)**:

Configure a Load Balancer, an ingress, and 2 services. Have service 2 talk to service one using a DNS name that maps to the Load Balancer.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T08:13:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.2-gke.1", GitCommit:"75d2af854b1df023c7ce10a8795b85d3dd1f8d37", GitTreeState:"clean", BuildDate:"2018-05-10T17:23:18Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
Google GKE

- OS (e.g. from /etc/os-release):
BUILD_ID=10452.89.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=e2e439017d740b3fbe0f4f1a2bc63af84facf535
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=66
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=66
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos

- Kernel (e.g. `uname -a`):
Linux gke-k8s-default-pool-d0b73bf2-d3ft 4.14.22+ #1 SMP Thu May 10 17:54:42 PDT 2018 x86_64 Intel(R) Xeon(R) CPU @ 2.00GHz GenuineIntel GNU/Linux

- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim <kiman@vmware.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
For any traffic to an LB IP that originates from the local node, re-route that traffic to the Kubernetes service chain. This allows traffic to an external LB from inside a cluster reachable. The implication of this is that internal traffic to an LB IP will need to go through SNAT. This is likely okay since source IP preservation with externalTrafficPolicy=Local only applies for external traffic anyways. The fix was spelled out in more detail by Tim here https://github.com/kubernetes/kubernetes/issues/65387. 

I think the correct behavior is to actually route the traffic to the LB instead of intercepting it with iptables but with the current set of rules I'm not sure this is possible. We also already have rules that route pods in the cluster cidr that want to reach LB IPs to the service chain:
```
-A KUBE-XLB-ECF5TUORC5E2ZCRD -s 10.8.0.0/14 -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -j KUBE-SVC-ECF5TUORC5E2ZCRD
```
Allowing traffic with `--src-type LOCAL` to do the same makes sense to me

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/65387 https://github.com/kubernetes/kubernetes/issues/66607

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
iptables proxier: route local traffic to LB IPs to service chain
```

### CODE REVIEW

**Comment by andrewsykim** [comment]
@m1093782566 @Lion-Wei any ideas if we need this for IPVS proxier?

**Comment by jcodybaker** [comment]
@andrewsykim - I've tested this and unfortunately it doesn't seemed to have changed the behavior seen in #66607.  I've left my test cluster up, and can provide any debug that's helpful--just message on slack.  I've started debugging by looking through the iptables counter.  I didn't get the full picture tonight, but it's does seem to be taking a different path than we were seeing before.

**Comment by andrewsykim** [comment]
@jcodybaker can you share the output for `sudo iptables-save` please?

**Comment by mauilion** [comment]
@andrewsykim 
In my testing I was able to reproduce this case:
Bring up a 3 node cluster and schedule pods on 2 of them.
Define a service of type loadbalancer and ensure that the loadbalancer provides an ip address
modify the service and set externalTrafficPolicy to Local
Bring up a pod with hostNetwork: true on the node where there are no endpoints.
before your change I am not able to connect to the service.
after your change I can.

**Comment by andrewsykim** [comment]
I spoke with @jcodybaker on slack. What he was looking for was the ability to preserve source IP (via proxy protocol from the LB) for local traffic in the cluster via the LB IP. I'm not sure this is possible or even supported for internal traffic. Internal traffic should use cluster IP and the externalTrafficPolicy type shouldn't matter in that case. 

With that said, it _might_ be possible to remove the fw chain that detects the LB IP as the destination and actually route traffic to the LB instead of intercepting it with itpables. This probably needs a longer discussion/proposal though.

**Comment by mauilion** [comment]
@andrewsykim Here is the article I put  together for testing this stuff:
https://mauilion.dev/posts/kind-k8s-testing/

Thanks!

**Comment by donbowman** [comment]
> I spoke with @jcodybaker on slack. What he was looking for was the ability to preserve source IP (via proxy protocol from the LB) for local traffic in the cluster via the LB IP. I'm not sure this is possible or even supported for internal traffic. Internal traffic should use cluster IP and the externalTrafficPolicy type shouldn't matter in that case.
> 
> With that said, it _might_ be possible to remove the fw chain that detects the LB IP as the destination and actually route traffic to the LB instead of intercepting it with itpables. This probably needs a longer discussion/proposal though.

We are in the process of upstreaming all the PR to do this w/ envoy + istio (it uses the sidecar to unpack the original public IP in the pod).
https://github.com/envoyproxy/envoy/issues/4128 discusses that.

**Comment by kinolaev** [comment]
Maybe we can just ACCEPT packets to loadBalancerIP instead of redirecting it to SVC chain? You can test it with:
```
sudo iptables -t nat -I ${XLB chain name} 2 -m addrtype --src-type LOCAL -j ACCEPT
wget -Y off -O- https://${LB IP}
sudo iptables -nvL --line-numbers -t nat | grep  -A 4 "Chain ${XLB chain name}"
```

**Comment by bretep** [comment]
@andrewsykim Thank you for the work on this. Just the behavior I was expecting and was surprised it didn't work that way. 👍 

@thockin Is this PR something you will be looking at in the near future? Trying to gauge if I should fork and build... I'd rather not, but I know everyone is busy. ;)

**Comment by kinolaev** [comment]
@andrewsykim, I think we need to route local traffic to `svcChain` only if there are no local endpoints. Because if node has local endpoint we need to route traffic only to local `endpointChain` regardless of the source of traffic (as it works now).
So I suggest:
1) move jump to `svcChain` to [L1231](https://github.com/kubernetes/kubernetes/blob/8dfd4def99ea1cc6296b25bc2121af9a72661fdf/pkg/proxy/iptables/proxier.go#L1231)
2) remove jump to `KubeMarkMasqChain` (I don't no why but it works without masquerading)

If you do this then my PR #78247 will be differ by only one thing: `ACCEPT` to send traffic to LB vs `svcChain` to send traffic to remote pod.

Using `ACCEPT` preserves source IP but requires network access from node to LB.
Using `svcChain` doesn't requires anything but source IP will be lost.

**Comment by andrewsykim** [comment]
@kinolaev I _think_ I agree with you but not sure yet. Traffic originating from the cluster cidr AND directed at the LB IP goes through the service chain if `externalTrafficPolicy=Local`. Shouldn't the same policy apply for nodes then? Either way, let me try to test your PR to make sure there's no edge cases missing there. I wonder if the `src-type` check can lead to interesting behavior if we go back out to the external LB and back into the node pool.

**Comment by kinolaev** [comment]
> Shouldn't the same policy apply for nodes then?
We can remove `src-type` check and jump to SVC chain for traffic from cluster CIDR to send all traffic from nodes that don’t have local endpoints to LB. This solution looks more consistent, but requires network access from node to LB for traffic from pods and has no benefit (source IP will be lost). So I don’t think that we need to choose this solution just for consistency.

**Comment by andrewsykim** [comment]
> So I don’t think that we need to choose this solution just for consistency.

Agreed that it shouldn't just be for consistency sake. The question I have boils down to: if we know the destination IPs for a given external LB IP, do we route back to external IP or route directly to the backend? We route to directly to service backend for the pod cidr case, wondering if there was a valid reason for that which also applies to nodes. Need to test all the scenarios first.

**Comment by kinolaev** [comment]
I think routing directly to backend is good solution (and personally I prefer it). But if maintainers will decide that source IP must be preserved they will already have a solution for this.

Please see my comment above about moving routing to SVC after check that node doesn't have local endpoints, because otherwise traffic must be routed locally. Do you agree?

**Comment by andrewsykim** [comment]
> Please see my comment above about moving routing to SVC after check that node doesn't have local endpoints, because otherwise traffic must be routed locally. Do you agree?

I'm not sure, need to put more thought into this because the existing local endpoint logic only applies for external traffic coming into the cluster (hence `externalTrafficPolicy=Local`). Preferring local endpoint if they exist and falling back to service chain _sounds right_ but I don't think it's that simple, we are possibly changing the expectations for internal traffic in ways that can be unexpected. This would be addressed by on-going work for service topology which I would say is out-of-scope for what we're trying to fix here. I'm not the authoritative source on this though :)

**Comment by kinolaev** [comment]
I see now the root of our discussion: we have different borders of "internal") For me "internal" means pods and services but you also include nodes to "internal". So for me nodes' outgoing traffic is "external" traffic and must be routed to local endpoints or load balancer. But if nodes' outgoing traffic is "internal" traffic then your solution is most logical.
@thockin, waiting for you)

**Comment by andrewsykim** [comment]
Err sorry, forgot to put a milestone on this one. Putting the milestone label so it's on the radar at least. It's a bit last minute so feel free to bump into next release (or cherry-pick after code freeze)

/milestone v1.15

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -1202,6 +1202,16 @@ func (proxier *Proxier) syncProxyRules() {
...
+		// Masqueraded traffic in this scenario is okay since source IP preservation only applies to external traffic anyways.
+		args = append(args[:0], "-A", string(svcXlbChain))
+		writeLine(proxier.natRules, append(args,
+			"-m", "comment", "--comment", fmt.Sprintf(`"masquerade LOCAL traffic for %s LB IP"`, svcNameString),
+			"-m", "addrtype", "--src-type", "LOCAL", "-j", string(KubeMarkMasqChain))...)
+		writeLine(proxier.natRules, append(args,
+			"-m", "comment", "--comment", fmt.Sprintf(`"route LOCAL traffic for %s LB IP to service chain"`, svcNameString),
+			"-m", "addrtype", "--src-type", "LOCAL", "-j", string(svcChain))...)
+
 		// First rule in the chain redirects all pod -> external VIP traffic to the
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Minor point - the comment here says "first rule", but you broke that :)

Can you put the new block after this block, and word the comments in similar fashion?

**Review by thockin — COMMENTED** [review]
Thanks.  I'll approve now, but can you please send a minor fixup?

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/proxy/iptables/proxier.go]: Will do, thanks Tim!

**Comment by snuxoll** [comment]
@andrewsykim @kinolaev It's not just preserving the source IP that is the issue, the current behavior for routing traffic to LoadBalancer services within the cluster makes a false assumption that the LoadBalancer is always transparent. In the case of DigitalOcean, as an example, this is not the case - the LoadBalancer supports the PROXY protocol as well as TLS termination - sending traffic directly to a pod will often result in unexpected behavior as a result.

This is only an issue when LoadBalancer's provide an IP address in the status spec since k8s does not attempt to magic the LoadBalancer away when a hostname is provided instead, as is the case with ELB's from AWS, for example. If your cloud provider supports network transparent load balancing than I don't think the optimization needs to be thrown away, but cloud controllers should have a way to indicate to the kublet that a LoadBalancer service is not transparent and shouldn't be routed internally.

**Pull request merged**